### ISSUE TITLE
SAML Response without Signature Node and Encrypted Assertion

### ISSUE THREAD

**Issue by HeikoLux** [Sustainsys/Saml2#449]
Hallo,
we use ADFS and this wonderful SP for SAML.
The authentication process went right but validation signature of response fails.
At this point in class XmlHelper

  var signatureElement = xmlElement["Signature", SignedXml.XmlDsigNamespaceUrl];
            if (signatureElement == null)
            {
                signatureElement = xmlElement["", SignedXml.XmlDsigNamespaceUrl];
            }

signatureElement is null. In response no Signature node appears. We sign Message and Assertion.
Can you help us?

/Heiko

**Comment by AndersAbel** [comment]
Can you show what the SAML message looks like? Use SAML Dev tools for Chrome for SAMLTracer for Firefox to capture it.

**Comment by HeikoLux** [comment]
`<samlp:Response 
    ID="_2676c66e-9369-432a-8743-20bdf12bb1e2" 
    Version="2.0" 
    IssueInstant="2016-04-16T07:57:34.041Z" 
    Destination="https://localhost:44328/AuthServices/Acs" 
    Consent="urn:oasis:names:tc:SAML:2.0:consent:unspecified" 
    InResponseTo="id88248af2a2d1454d9ac4ba47d0cd44ae" 
    xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol">
    <Issuer 
        xmlns="urn:oasis:names:tc:SAML:2.0:assertion">XXXX
    </Issuer>
    <samlp:Status>
        <samlp:StatusCode 
            Value="urn:oasis:names:tc:SAML:2.0:status:Success" />
    </samlp:Status>
    <EncryptedAssertion 
        xmlns="urn:oasis:names:tc:SAML:2.0:assertion">
        <xenc:EncryptedData 
            Type="http://www.w3.org/2001/04/xmlenc#Element" 
            xmlns:xenc="http://www.w3.org/2001/04/xmlenc#">
            <xenc:EncryptionMethod 
                Algorithm="http://www.w3.org/2001/04/xmlenc#aes256-cbc" />
            <KeyInfo 
                xmlns="http://www.w3.org/2000/09/xmldsig#">
                <e:EncryptedKey 
                    xmlns:e="http://www.w3.org/2001/04/xmlenc#">
                    <e:EncryptionMethod 
                        Algorithm="http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p">
                        <DigestMethod 
                            Algorithm="http://www.w3.org/2000/09/xmldsig#sha1" />
                    </e:EncryptionMethod>
                    <KeyInfo>
                        <ds:X509Data 
                            xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
                            <ds:X509IssuerSerial>
                                <ds:X509IssuerName>
                                    CN=e-ADFS-Sub1-CA, 
                                    DC=e-adfs, 
                                    DC=lew, 
                                    DC=de
                                </ds:X509IssuerName>
                                <ds:X509SerialNumber>1984766322757178865934056788816590107140161552</ds:X509SerialNumber>
                            </ds:X509IssuerSerial>
                        </ds:X509Data>
                    </KeyInfo>
                    <e:CipherData>
                        <e:CipherValue>CQXoRvO7UqxMtVf87gHcSpJ3ztf0xlv2SuylQspJVUx9aeuy2TZ9wEAzOaw/DDtt3TvMB2I0uPfFfdN0r8BxhRFwXKEORJWOBVNUSZciw9BYVbCeUHv8PB7CN3rKTvWqKm2TBkpAHsHrI8d7dh/F2vI/qh8B4PNagGmCBTbMo5UIVvPM1SB3U8d60a7QWnBo0efn2Ecv3ZmXeLWu9kWagg5ORtY+c6gs7iuszTcQQiOHBIvY9uqSQaFdMPAD6kyC1E0MIaHK/cSupbN0IaFzW/Moc/
                            mn7oPBbGSZ37t82qFJh1OnJuEaDSsZh4mnXffZuuFRXeHPJkpSEvrDc3JR4g==
                        </e:CipherValue>
                    </e:CipherData>
                </e:EncryptedKey>
            </KeyInfo>
            <xenc:CipherData>
                <xenc:CipherValue>yAxaX12kHr73ednBXEPjNohAbrcUjVbVEYOVz8URB/m8i9E6spnMWPFdyxGt5ZShEXnzW26sLoEz5NHCbIIjMETMVfU+4tLUujZn4noS69iEE7SW3pGXEMkzvdPLZzYXmis3vQLTIC2NDocz3T9QhpXspExYLoTeu1hz8pL8nzbMc/pQFVYM3I2UI0LJBAeFUmR7e2qtnogXLW3ykvmRXrOhP7P74XOfe3+z+Y4YeOlEmgix9NVxOXCaTdy6vfxBOrgrcgF25mcm6XlP3WV1BsyCLxFF6HmJreT8Lrw9vuiE6cf0mjTxUVj/ljIB5pv/MeZZC4Fe+LZQfPvlGYwai/ntDfcg3IPGPrWCgdrq/wwmTzaTTKj5D+wdUEptNHoQWQXFAuoLFTELptWlC4M8G96OeYaLU8n1czAMlDSfF2fq7A8nEMrTBO44cQ0Q+gK0rcL4Kn1UJhfmJdsbptRFqALkb/wzEkz2KLOtTxPaqjbYULPFddONomvA1QX05MwmX4ZjxoF0VRi6ElMThJBxjxlgriOES9wjPMwlyv297WOTwlL1NjxOqmbAt9R+VFdZP4/3jw8+KsCYtWmbD8ury3CdRp5y/D7bApae+ULDsdT9twPrudVt9/Q0yIT8pnpBZ2sDY40OSj0rsXH+813qdBX6YlReh+GZvAeoO6mwlErxL9ytvq0p9Q+yZag1cogbHQIv3yu6IOUwtpaMdTlHyhIwQxaDibZfgCfPRgQKAJ8vyEUQtd6yLxrWDb/4Vg0LzsUIKO12r4WTzLWQ1Io0fwq4SGLMjZlhdkUj1Fe3B0yOqMTYZV4u9EMviry/5Z0NV1/EHCD583cAt8Q9uVzhuu/A6vvxOkCgogKUVmGn3hu3VaUKePLvlceDfC1lHB3nPIn8Xqgx7JlJiwijTLFW5BFqeksM26BB4ne9Ya3gXdBzBb1WLVxNvdiRzYzy/lNk8OujVsibbMuzpBh43phfeyB6XMm93Zk/IJnsktYPWfEI7ZMKA2plS8KdMus3w3ZRJ3ohmqQpjXIM7nEJOlUol49glASOfNohgLOMaxDAxTvzaUSJFbsZYNQ0RiTLYxwh/8S5Qb3i4FMH80WLx510QpK12kgKNlLfmZMo6GeBGAl4Q4t1I7i1g9J/vLU+PX6toIBET7iyouWog6OYjYd9zZPjKHnCF7ZFjfdmsTGpRzUx6ldLlrr8BV8dQRLPrcLHcEdpJggOQMdLrlmDQu/FW9JzB0t++lOAQfAFteetHFcJb0/slhznsE6tSJRg+Fv9j3vEh7jaog8GIJQtHOEMacrGcZLj3uuEWrjhctv4bmOsRqWTemSueo+7jhjSA/J4kdpXQBJ1RBwo6mnbR2icItaVm3dUNmB0/rIUEhUHUIlgVBdEJ1qKuP45W9dFkdSeZ/KyA1rdbGu+bFytPlpFr3bQtWrCXDg4hy0GsMoDWNRjCpvCFM/6WjKn1luK6DmwDhGch4n57wzuraKMFbFJL2Y3lJz9kog/n6e2fKJqfX3A+8R5c/O+/4VS6gy9UEkUGFCI0ooIoMereT5soNPfYAOIq65Fp+TJ/
                    uCczUAJR7Q=
                </xenc:CipherValue>
            </xenc:CipherData>
        </xenc:EncryptedData>
    </EncryptedAssertion>
</samlp:Response>`

This is the response. There is no Signature node. But on ADFS Message and Assertion should be with signature

**Comment by AndersAbel** [comment]
If the signature is on the assertion node, it is part of the encrypted data. I think that we check the signature before decrypting the assertions. We should probably do it the other way around. For now, a workaround is to disable encrypted assertions.

**Comment by HeikoLux** [comment]
Signature MessageOnly would be fine?

**Comment by AndersAbel** [comment]
If you can configure ADFS to add a signature to the message that will work too. The signature on the assertion is ignored in that case.

**Comment by HeikoLux** [comment]
On ADFS I configured Endpoint with POST Binding and everything worked, but with a little workaround on my own. When signature is added in class Saml2RedirectBinding, I coded sha256 hash string on myself (with sha256 support in Global.asax GlobalEnableSha256XmlSignatures):
Original code
`    queryString += "&SigAlg=" + Uri.EscapeDataString(SignedXml.XmlDsigRSASHA1Url);

            var signatureDescription = 
                (SignatureDescription)CryptoConfig.CreateFromName(SignedXml.XmlDsigRSASHA1Url);`

My workaround
`         queryString += "&SigAlg=" + Uri.EscapeDataString("http://www.w3.org/2001/04/xmldsig-more#rsa-sha256");

            var signatureDescription =
                (SignatureDescription)CryptoConfig.CreateFromName("http://www.w3.org/2001/04/xmldsig-more#rsa-sha256");`

Without this, it is not possible to verify Signature because SHA1 is used.
What can I do without modifying code?

/Heiko
